FEAT Support precomputed distance matrix for `PairwiseDistancesReductions`

kyrajeep opened the pull request:
Towards #25888

1. What does this implement/fix? Explain your changes.
   (1) To take in precomputed distance directly, uses 'compute' in {ArgKmin, RadiusNeighbors}{ClassMode} {32,64}.
  (2) When this is received as a parameter X, initializes an instance of class PrecomputedDistance.
  (3) This precomputed data is formatted by datasets_pair.py
  (4) The data is passed onto the backend by using the get_for() factory method.
  (5) Changes the API according to @Micky774 's suggestion, which is to include 'precomputed' as one of the metrics instead of as an additional parameter. This simplifies its use by taking a precomputed matrix as X and avoids confusion for the users who do not have a precomputed matrix. 
    For example, a user with a precomputed matrix specifies metric = 'precomputed' and X = precomputed matrix. Previously, there was an extra 'precomputed' parameter that took a Boolean value.

2. Any other comments?
Thank you to @jjerphan for the reviews!

Co-authors: @jjerphan @adam2392

adam2392 committed:
Co-authored-by: @jjerphan Add initial commit

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'main' into precomputed

kyrajeep committed:
Merge branch 'main' into precomputed
This feature was implemented last year so updates need to be pulled

kyrajeep committed:
Add precomputed option in dispatcher classes

kyrajeep committed:
comments, questions, code to feature precomputed and maintain class hierarchy

kyrajeep committed:
Merge branch 'scikit-learn:main' into feat_precomputed

kyrajeep committed:
Add the precomputed option in argkmin_classmode

kyrajeep committed:
Merge remote-tracking branch 'refs/remotes/origin/feat_precomputed' into feat_precomputed

kyrajeep committed:
Update is_usable_for to use the dispatcher for precomputed

kyrajeep committed:
Changed the name from distance_matrix to precomputed_distance

Micky774 commented:
> Also, in the abstract methods dist and surrogate_dist of PrecomputedDistanceMatrix, the indices [i,j] are required. Does this mean that other indices in the same dataset need computation and are not precomputed?

No. Keep in mind that `surrogate_dist` and `dist` methods are for scalar distances, i.e. they take a pair of indices and produce the distance between exactly those two data vectors. Hence, for the precomputed distances, it's simply indexing into the precomputed distances array.

kyrajeep commented:
Thank you for the clarification

kyrajeep commented:
In class PrecomputedDistanceMatrix, I don't see the option to input the precomputed distance matrix. Should I make that option or is it stored somewhere else?

jjerphan commented:
Hi!

A reference to the distance matrix needs to be passed to the constructors of `PairwiseDistancesReductions` subclasses (as it is the case with `X` and `Y`) and be stored in the base cython class.

Some logic with the chucking before the computations also needs adaptation.

We could have another second class hierarchy for handling precomputed distance matrices, but let's start with this simpler adaptation for now.

kyrajeep committed:
Start of changing BaseDistancesReduction superclass

kyrajeep committed:
Add the function to return the precomputed matrix

kyrajeep committed:
Merge branch 'scikit-learn:main' into feat_precomputed

kyrajeep committed:
Merge branch 'scikit-learn:main' into feat_precomputed

kyrajeep committed:
Add to BaseDistanceReduction the function to take in and return a precomputed matrix

kyrajeep committed:
Add comments to work on the subclass of DatasetsPair

kyrajeep committed:
Alter the compute method of ArgKminClassMode to take in precomputed matrix stored in datasets_pair.pyx

kyrajeep commented:
Thank you for your comments. They make sense, are helpful, and I believe I am going in the right direction. So far I have made progress on all the points made above by @jjerphan except for creating an instance to store the precomputed matrix. I think even though there are things (referencing the precomputed matrix and syntax) that should be definitely changed a review will be helpful and probably worthwhile at this point :) Higher level review will be helpful enough as I am familiar with the class hierarchy and codebase.

I have a couple questions that are not syntax related. We have a class for precomputed that subclasses DatasetsPair. In that class we have surrogate_dist and dist functions. My current understanding is that they are both used to compute distance depending on metric, chunking strategy, etc.. I want to create a method that will take in a precomputed matrix, if provided, and another method to retrieve this matrix to be passed into the compute method. Does that sound about right?

As for creating an instance to store the precomputed matrix itself... which file should I include it in?

Feedback will be appreciated! Thank you!

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx.tp:
```diff
@@ -158,6 +164,55 @@ cdef class DatasetsPair{{name_suffix}}:
+
+    Parameters
+    ----------
+    Must receive precomputed_distance: ndarray of shape (n_samples_X, n_samples_Y)? Really? 
+        Must be C-contiguous. 
```
> Really?

IMO, it would make sens to support C-contiguity for now. We can always extend it then.

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx.tp:
```diff
@@ -158,6 +164,55 @@ cdef class DatasetsPair{{name_suffix}}:
+
+    def __init__(
+        self,
+        const {{INPUT_DTYPE_t}}[:, ::1] precomputed_distance,  # do we break it up into 64 and 32?
```
I would always use `float64` even for `float32` datasets, at least at first.

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx.tp:
```diff
@@ -158,6 +164,55 @@ cdef class DatasetsPair{{name_suffix}}:
+    ):
+        super().__init__(
+            distance_metric={{DistanceMetric}}(),
+            n_features=0,  #what is n_features? aren't there more we need to take in?
```
We do not need `n_feature` for this case, so we can ignore it, but let's pass the actual value for consistency at the constructors' call sites.

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx.tp:
```diff
@@ -158,6 +164,55 @@ cdef class DatasetsPair{{name_suffix}}:
+      this is a surrogate distance that is used for real computations
+  
+
+    '''
+    # this is a surrogate distance that is used for real computations
+    # may not need to specify since this subclass does not recompute
+    @final
+    cdef DTYPE_t surrogate_dist(self, ITYPE_t i, ITYPE_t j) noexcept nogil:
+        return self.precomputed_distance[i, j]
+    '''
```
We need it to respect the base class' interface.
```suggestion
    @final
    cdef DTYPE_t surrogate_dist(self, ITYPE_t i, ITYPE_t j) noexcept nogil:
        return self.precomputed_distance[i, j]

```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx.tp:
```diff
@@ -158,6 +164,55 @@ cdef class DatasetsPair{{name_suffix}}:
+    cdef DTYPE_t surrogate_dist(self, ITYPE_t i, ITYPE_t j) noexcept nogil:
+        return self.precomputed_distance[i, j]
+    '''
+    # i don't believe this needs to be indexed.. but needs to be defined as 
+    # it is an abstract method
+    @final
+    cdef DTYPE_t dist(self) noexcept nogil:
+        return self.precomputed_distance
```
```suggestion
    @final
    cdef DTYPE_t dist(self, ITYPE_t i, ITYPE_t j) noexcept nogil:
        return self.precomputed_distance[i, j]

```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py:
```diff
@@ -192,6 +198,7 @@ def compute(
         Y,
         k,
         metric="euclidean",
+        precomputed_matrix = None,
```
Can you move it bellow `X` and `Y`, make their default value `None` and check in the implementation that `X == Y == None` xor `precomputed_matrix == None`?

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py:
```diff
@@ -421,6 +428,10 @@ def compute(
         for the concrete implementation are therefore freed when this classmethod
         returns.
         """
+        # TODO: to maintain RAII, look at the implementation of the compute method
+        # if metric == 'precomputed':
+        #   return PrecomputedDistanceMatrix.precomputed_distance()
```
`DatasetsPair.get_for` must be adapted accordingly to return an instance of the new class `PrecomputedDistanceMatrix{32,64}` you have created when `precomputed_matrix != None`.

Then, the precomputed distance matrix simply need to be forward to the implementation `compute` method and be pass to the `DatasetsPair.get_for` class method (e.g. see [this line for `ArgKmin`](https://github.com/scikit-learn/scikit-learn/blob/a97d584196da14084b33b71d80712451e26a5b0a/sklearn/metrics/_pairwise_distances_reduction/_argkmin.pyx.tp#90). 

Similar changes must be made to the other class (e.g. `{ArgKmin,RadiusNeighbors}{ClassMode,}{32,64}`, excepted the ones for the Euclidean specialization (e.g. `Euclidean{ArgKmin,RadiusNeighbors}{32,64}`).

jjerphan commented on sklearn/metrics/_dist_metrics.pyx.tp:
This file must not be changed (this PR must not deal with any `DistanceMetric`'s implementations).

jjerphan reviewed (commented):
A quick review with more detailed explanations than my previous ones.

I pushed a commit (see 3a027c7) since your local `venv_sklearn` was staged and committed as part of e2ef246. I would recommend updating your global `.gitignore` with `venv*` to avoid this in the future, or to use a conda environment.

jjerphan committed:
Delete `venv_sklearn`

Signed-off-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep committed:
Delete a line regarding precomputed since this file is not to be changed

kyrajeep committed:
Merge remote-tracking branch 'refs/remotes/origin/feat_precomputed' into feat_precomputed

kyrajeep committed:
Merge branch 'scikit-learn:main' into feat_precomputed

kyrajeep committed:
Update sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx.tp

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep committed:
Add venv/ to ignore my local virtual environment

kyrajeep committed:
Merge branch 'feat_precomputed' of https://github.com/kyrajeep/scikit-learn into feat_precomputed

Alter the code according to guidelines and resolve conflicts

kyrajeep committed:
Write the methods to fulfill the abstract class requirements

kyrajeep committed:
Set default for input data to none and check if provided correctly for ArgKmin

kyrajeep committed:
Update the XOR check

kyrajeep committed:
Modify get_for classmethod

kyrajeep committed:
Modify get_for classmethod

kyrajeep committed:
Merge branch 'scikit-learn:main' into feat_precomputed

jjerphan commented:
Is everything alright, @kyrajeep?

kyrajeep commented:
> Is everything alright, @kyrajeep?

Hi! Thank you for checking in. I made most of the changes from your review which was very helpful. I marked them as resolved. I do need to look at it again and there is a point I want to ask you. I will be working on it this weekend!

kyrajeep committed:
Modify dispatcher's {ArgKmin, RadiusNeighbors}{ClassMode}{32,64} for the input, precomputed

kyrajeep committed:
fixed by the linter black

kyrajeep commented:
I believe I finally addressed all the comments from the review! Thank you so much for the review @jjerphan. I have been working on the code little by little to address each comment. It took much longer than expected but learned a lot.

I was confused about how the computation and receiving input are decoupled but I understand now. Correct me if this is wrong: 

ArgKmin in dispatcher.py receives the input, but ArgKminClassMode in dispatcher.py does all the computational heavy lifting. ArgKmin{32,64} and ArgKminClassMode{32,64} only inherit from ArgKmin in dispatcher.py, making it separate from the computation part. The same goes for RadiusNeighbors.

jjerphan reviewed (changes requested):
Your remark is not entirely exact.

For an overview of the architecture with dispatcher and implementations, see:

https://github.com/scikit-learn/scikit-learn/blob/ede02b4d000d3f9334940cf5c81155f9dc070e77/sklearn/metrics/_pairwise_distances_reduction/__init__.py#L44-L94

Doc-strings give more information about each individual components.

Tests should also be added for the various `PairwiseDistancesReductions`.

kyrajeep committed:
Update sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx.tp

Edit the comment

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep committed:
Correct the syntax as requested

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep committed:
Remove default value of none

because that was done in its base class

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep commented on sklearn/metrics/_pairwise_distances_reduction/_argkmin.pyx.tp:
```diff
@@ -56,6 +57,7 @@ cdef class ArgKmin{{name_suffix}}(BaseDistancesReduction{{name_suffix}}):
 
         No instance should directly be created outside of this class method.
         """
+            
```
Yes! I thought the linter black got rid of it. Is there another linter I should be using or should I manually do it? Thank you.

kyrajeep committed:
Remove blank lines..

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep committed:
Update pyproject.toml

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep committed:
Update the doctring

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep committed:
Remove blank lines

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep committed:
Remove a questions in comment

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep committed:
Update .gitignore to not commit a virtual env

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep committed:
Remove unnecessary blanks

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

jjerphan commented on .gitignore:
```diff
@@ -44,6 +44,7 @@ coverages
 samples
 doc/coverages
 doc/samples
+venv/
```
I would recommend using [miniforge](https://github.com/conda-forge/miniforge) with [conda-forge](https://conda-forge.org/).

kyrajeep commented:
Hello @jjerphan ,

My availability got very limited the last couple months (hence the little communication) but I have some time today and am learning how to use pytest. I will let you know when I am done! Thanks!

jjerphan commented:
Hi @kyrajeep,

There's no problem, take your time.

kyrajeep commented:
Hello! I got back from our yearly vacation and am working on the tests for the precomputed feature. In test_pairwise_distance_reduction.py, I created a function to check the shape of the precomputed matrix. I was wondering where the functions in test_pairwise_distance_reduction are called.

kyrajeep committed:
Pass the precomuted matrix as a DatasetsPair object directly

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep committed:
Initial commit to check the precomputed input array size

kyrajeep committed:
Added tests to check for NaNs, np.array, data types

kyrajeep committed:
Revert some unnecessary type casting

kyrajeep committed:
Fixing minor issues

kyrajeep committed:
Merge branch 'scikit-learn:main' into feat_precomputed

kyrajeep committed:
Fix the tests

kyrajeep committed:
bug fixes regarding data types, variable declaration with cython

kyrajeep committed:
Merge branch 'scikit-learn:main' into feat_precomputed

jjerphan commented:
You have committed your virtual environment. 

Can you delete it as previously explained?

kyrajeep commented:
> You have committed your virtual environment.
> 
> Can you delete it as previously explained?

Oh, that's quite surprising as I checked .gitignore before committing and even added more lines like *venv* and *venv/. I even git added the files that are venv files specifically to avoid this issue. My apologies.

kyrajeep committed:
Delete all venv files

kyrajeep committed:
Merge branch 'feat_precomputed' of https://github.com/kyrajeep/scikit-learn into feat_precomputed

kyrajeep committed:
Delete venv_sklearn directory

jjerphan commented on .gitignore:
Can you revert the changes made to this file?

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_argkmin_classmode.pyx.tp:
```diff
@@ -54,8 +56,9 @@ cdef class ArgKminClassMode{{name_suffix}}(ArgKmin{{name_suffix}}):
         """
         # Use a generic implementation that handles most scipy
         # metrics by computing the distances between 2 vectors at a time.
+
```
```suggestion
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_base.pxd.tp:
Can you revert the changes made to this file?

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_base.pyx.tp:
```diff
@@ -135,9 +138,10 @@ cdef class BaseDistancesReduction{{name_suffix}}:
         cdef:
             intp_t X_n_full_chunks, Y_n_full_chunks
-
+       
```
```suggestion

```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_base.pyx.tp:
```diff
@@ -135,9 +138,10 @@ cdef class BaseDistancesReduction{{name_suffix}}:
         chunk_size=None,
         strategy=None,
      ):
+
```
```suggestion
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx.tp:
```diff
@@ -158,6 +165,47 @@ cdef class DatasetsPair{{name_suffix}}:
+
+    @final
+    cdef intp_t n_samples_X(self) noexcept nogil:
+        with gil:
```
You must not take the gil here and in the other methods.

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py:
```diff
@@ -81,10 +81,12 @@ def valid_metrics(cls) -> List[str]:
 
     @classmethod
-    def is_usable_for(cls, X, Y, metric) -> bool:
+    def is_usable_for(cls, metric, X = None, Y = None, precomputed = None) -> bool:
```
You must not change the order here.

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py:
```diff
@@ -105,6 +109,12 @@ def is_usable_for(cls, X, Y, metric) -> bool:
         True if the dispatcher can be used, else False.
         """
+        if precomputed is not None or (X is not None and Y is not None):
+            print("input is valid")
```
Can you remove this branch and negate the condition just above?

jjerphan commented on sklearn/metrics/tests/test_pairwise_distances_reduction.py:
```diff
@@ -485,10 +552,9 @@ def test_assert_compatible_argkmin_results():
+        
 @pytest.mark.parametrize("check_sorted", [True, False])
-def test_assert_compatible_radius_results(check_sorted):
+def test_assert_compatible_radius_results(check_sorted: bool):
```
scikit-learn does not use type annotations at all. Can you remove all of them? Thank you.

jjerphan commented on sklearn/metrics/tests/test_pairwise_distances_reduction.py:
```diff
@@ -347,6 +388,32 @@ def assert_compatible_radius_results(
+    # Success Case: Valid precomputed matrix
+    n_samples_X, n_samples_Y = 5, 5
+    valid_precomputed = np.random.rand(n_samples_X, n_samples_Y).astype(np.float32)
+    try:
```
This `try`-`except` has to be removed.

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py:
```diff
@@ -421,6 +446,17 @@ def compute(
         for the concrete implementation are therefore freed when this classmethod
         returns.
         """
+        if X is None and Y is None and precomputed is None:
```
Can you add tests for all of those `ValueError`?

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py:
```diff
@@ -277,10 +288,22 @@ def compute(
         for the concrete implementation are therefore freed when this classmethod
         returns.
         """
+        if X is None and Y is None and precomputed_matrix is None:
```
Can you add tests for those `ValueError`?

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx.tp:
```diff
@@ -158,6 +165,47 @@ cdef class DatasetsPair{{name_suffix}}:
+            n_features=0,  
+        )
+        # This array has already been checked.
+        self.precomputed_distance = precomputed_distance
```
This explains why you needed to take the GIL.
```suggestion
        self.distance_matrix = precomputed_distance
```

kyrajeep replied:
Thanks.

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx.tp:
```diff
@@ -55,6 +55,7 @@ cdef class DatasetsPair{{name_suffix}}:
         cls,
         X,
```
`X` and `Y` might now be `None` as well.

kyrajeep replied:
Yes, thank you.

jjerphan reviewed (commented):
Continuing reviews.

kyrajeep committed:
Merge branch 'feat_precomputed' of https://github.com/kyrajeep/scikit-learn into feat_precomputed

kyrajeep committed:
Delete a blank line

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep committed:
Fix formatting

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep committed:
Fix formatting

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep committed:
Remove with gil for precomputed matrix

kyrajeep committed:
Merge branch 'feat_precomputed' of https://github.com/kyrajeep/scikit-learn into feat_precomputed
Pull changes made in the upstream

kyrajeep committed:
Fix formatting

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep committed:
Add tests for ValueErrors and fix the XOR for the method, is_usable.

kyrajeep committed:
Merge branch 'feat_precomputed' of https://github.com/kyrajeep/scikit-learn into feat_precomputed

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx.tp:
```diff
@@ -158,6 +165,43 @@ cdef class DatasetsPair{{name_suffix}}:
+
+    @final
+    cdef intp_t n_samples_X(self) noexcept nogil:
+        return self.precomputed_distance.shape[0]
```
```suggestion
        return self.distance_matrix.shape[0]
```

Same with the lines bellow.

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx.tp:
```diff
@@ -158,6 +165,43 @@ cdef class DatasetsPair{{name_suffix}}:
+
+    @final
+    cdef intp_t n_samples_Y(self) noexcept nogil:
+        return self.precomputed_distance.shape[1]
```
```suggestion
        return self.distance_matrix.shape[1]
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx.tp:
```diff
@@ -158,6 +165,43 @@ cdef class DatasetsPair{{name_suffix}}:
+  
+    @final
+    cdef float64_t surrogate_dist(self, intp_t i, intp_t j) noexcept nogil:
+        return self.precomputed_distance[i, j]
```
```suggestion
        return self.distance_matrix[i, j]
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx.tp:
```diff
@@ -158,6 +165,43 @@ cdef class DatasetsPair{{name_suffix}}:
+   
+    @final
+    cdef float64_t dist(self, intp_t i, intp_t j) noexcept nogil:
+        return self.precomputed_distance[i, j]
```
```suggestion
        return self.distance_matrix[i, j]
```

jjerphan commented on sklearn/metrics/tests/test_pairwise_distances_reduction.py:
```diff
@@ -1120,8 +1210,8 @@ def test_radius_neighbors_classmode_factory_method_wrong_usages():
     global_random_seed,
-    Dispatcher,
-    dtype,
+    Dispatcher: ArgKmin | RadiusNeighbors,
```
Please remove all type annotations and spurious formating changes which have been introduced in this file (and elsewhere if any).

jjerphan reviewed (commented):
We are nearly there! :)

kyrajeep committed:
syntax fix

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep committed:
syntax fix

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep committed:
syntax fix

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep committed:
syntax fix

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kyrajeep committed:
Remove type annotations

kyrajeep commented:
> We are nearly there! :)

Thank you for the reviews!

kyrajeep committed:
Merge branch 'scikit-learn:main' into feat_precomputed

kyrajeep committed:
Merge branch 'scikit-learn:main' into feat_precomputed

jjerphan commented on .gitignore:
Please, revert changes to this file.

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_base.pxd.tp:
Please, revert changes to this file.

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_base.pyx.tp:
```diff
@@ -137,7 +140,6 @@ cdef class BaseDistancesReduction{{name_suffix}}:
      ):
         cdef:
             intp_t X_n_full_chunks, Y_n_full_chunks
-
```
Please remove this line deletion.

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py:
```diff
@@ -105,7 +109,9 @@ def is_usable_for(cls, X, Y, metric) -> bool:
         True if the dispatcher can be used, else False.
         """
-
+        is_usable = (bool(X) and bool(Y)) ^ bool(precomputed)
```
This might require formatting.
```suggestion
        is_usable = (X is not None and Y is not None) ^ bool(precomputed)
```

jjerphan commented on sklearn/metrics/tests/test_pairwise_distances_reduction.py:
We also need tests checking that the results of `PairwiseDistancesReduction` obtained respectively on `(X, Y)` or on their precomputed distance matrix are identical.

jjerphan reviewed (changes requested):
Please, run the linters and remove the unused fixture in the tests.

Micky774 commented:
Catching up on this PR. I wanted to mention that I don't know if I like the current API with `precomputed` as a separate argument, since usually we include it as an option for `metric`. I think keeping with tradition leads to a more consistent experience since e.g. you don't have a situation where both `precomputed` and `metric` are defined (which would be entirely redundant) and instead succinctly have `metric='precomputed'`.

To enable this, we could simply enforce an API where `metric='precomputed'` expects `X` to be the precomputed distances, and raises an error if `Y` is provided (in the rare case that folks accidentally use precomputed when they didn't mean to).

What do you all think?

jjerphan commented:
Thank you for shiming in, @Micky774.

We can indeed revise the API to get them in line with the public ones.

kyrajeep commented:
> Please, run the linters and remove the unused fixture in the tests.

Will do.

kyrajeep committed:
Revert change

kyrajeep commented:
> Catching up on this PR. I wanted to mention that I don't know if I like the current API with `precomputed` as a separate argument, since usually we include it as an option for `metric`. I think keeping with tradition leads to a more consistent experience since e.g. you don't have a situation where both `precomputed` and `metric` are defined (which would be entirely redundant) and instead succinctly have `metric='precomputed'`.
> 
> To enable this, we could simply enforce an API where `metric='precomputed'` expects `X` to be the precomputed distances, and raises an error if `Y` is provided (in the rare case that folks accidentally use precomputed when they didn't mean to).
> 
> What do you all think?

That sounds good. I may not have time to change it with this PR though.

kyrajeep committed:
Revert change

kyrajeep committed:
Merge branch 'feat_precomputed' of https://github.com/kyrajeep/scikit-learn into feat_precomputed

kyrajeep commented:
Hi @Micky774, 

I'm sorry to take a while to respond. That's a great point. I agree that including 'precomputed' in one of the options for metrics is clearer. How about making the change in a separate PR so that this PR does not get too lengthy? Thanks!

kyrajeep committed:
Merge branch 'main' into feat_precomputed

Micky774 commented:
> Hi @Micky774,
> 
> I'm sorry to take a while to respond. That's a great point. I agree that including 'precomputed' in one of the options for metrics is clearer. How about making the change in a separate PR so that this PR does not get too lengthy? Thanks!

Considering that it is a fairly key part of the API being affected by this PR, I think it is better to resolve it here. In my opinion it costs less to fix it now, than it would to start up a new PR for that change. And no worries on the delay -- I appreciate you continuing your work here!

kyrajeep commented:
Ok, thank you :)

kyrajeep committed:
Test the precomputed input against the actual computation for a sample

kyrajeep committed:
Add "precomputed" as one of the options for "metric"

kyrajeep commented:
Hi @jjerphan, @Micky774 

While working on implementing the change in the API to enable 'precomputed' as one of the metrics, I realized that X = precomputed argument may lead to issues due to the dimensions being different from X. This is from the current DatasetsPair docstring:

 X : {ndarray, sparse matrix} of shape (n_samples_X, n_features)
     Input data.
     If provided as a ndarray, it must be C-contiguous.
     If provided as a sparse matrix, it must be in CSR format.
 Y : {ndarray, sparse matrix} of shape (n_samples_Y, n_features)
     Input data.
     If provided as a ndarray, it must be C-contiguous.
     If provided as a sparse matrix, it must be in CSR format.
 precomputed : ndarray of shape (n_samples_X, n_samples_Y) 
     Precomputed distance if provided.

What do you all think?

Micky774 commented:
> Hi @jjerphan, @Micky774
> 
> While working on implementing the change in the API to enable 'precomputed' as one of the metrics, I realized that X = precomputed argument may lead to issues due to the dimensions being different from X. This is from the current DatasetsPair docstring:
> 
> X : {ndarray, sparse matrix} of shape (n_samples_X, n_features) Input data. If provided as a ndarray, it must be C-contiguous. If provided as a sparse matrix, it must be in CSR format. Y : {ndarray, sparse matrix} of shape (n_samples_Y, n_features) Input data. If provided as a ndarray, it must be C-contiguous. If provided as a sparse matrix, it must be in CSR format. precomputed : ndarray of shape (n_samples_X, n_samples_Y) Precomputed distance if provided.
> 
> What do you all think?

I'm not sure I see the problem here. If you're worried about the fact that an `NxN` pre-computed matrix may be a result of `NxD` data for arbitrary `D`, then I do not believe it is a problem since the responsibility lies on the user to ensure that they are entering the _correct_ precomputed matrix for their data -- that is not something we could (or should) verify imo.

If that's not what your concern is, then I may need a bit more of an explanation :)

kyrajeep commented:
> > Hi @jjerphan, @Micky774
> > While working on implementing the change in the API to enable 'precomputed' as one of the metrics, I realized that X = precomputed argument may lead to issues due to the dimensions being different from X. This is from the current DatasetsPair docstring:
> > X : {ndarray, sparse matrix} of shape (n_samples_X, n_features) Input data. If provided as a ndarray, it must be C-contiguous. If provided as a sparse matrix, it must be in CSR format. Y : {ndarray, sparse matrix} of shape (n_samples_Y, n_features) Input data. If provided as a ndarray, it must be C-contiguous. If provided as a sparse matrix, it must be in CSR format. precomputed : ndarray of shape (n_samples_X, n_samples_Y) Precomputed distance if provided.
> > What do you all think?
> 
> I'm not sure I see the problem here. If you're worried about the fact that an `NxN` pre-computed matrix may be a result of `NxD` data for arbitrary `D`, then I do not believe it is a problem since the responsibility lies on the user to ensure that they are entering the _correct_ precomputed matrix for their data -- that is not something we could (or should) verify imo.
> 
> If that's not what your concern is, then I may need a bit more of an explanation :)

Thank you for your input! That's exactly what I meant, but shouldn't we verify the shape of the precomputed matrix to throw an error if the dimension is not correct?

jjerphan commented:
I would look at and implement the same logic as the one of estimator taking a data matrix or a pre-computed distance matrix for `X`.

kyrajeep commented:
Sounds good. Thank you.

kyrajeep committed:
Change DatasetsPair to use the metric option to indicate precomputed

kyrajeep committed:
API change to include precomputed as a metric in the base abstract class

kyrajeep committed:
API change to include precomputed as one of the metrics: check X and Y

kyrajeep committed:
Change API to enable precomputed as a metric and take in X=precomputed_matrix

Alvaro-Kothe commented on sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py:
```diff
@@ -105,7 +109,15 @@ def is_usable_for(cls, X, Y, metric) -> bool:
+                is_usable = False
+
+        # is_usable = (X is not None and Y is not None) ^ bool(precomputed)
+        if is_usable == False:
```
Won't `is_usable` be undefined if `metric != "precomputed"`?

kyrajeep committed:
Update tests with the API change

kyrajeep committed:
Merge branch 'main' into feat_precomputed

kyrajeep committed:
Merge branch 'main' into feat_precomputed

kyrajeep committed:
Merge branch 'main' into feat_precomputed

jjerphan commented:
My two cents briefly is that you likely need to install and use `c-compiler` and `cxx-compiler` from `conda-forge` because Apple Clang does not support OpenMP.
